lib/rust/parser/src/syntax/operator.rs

```diff
@@ -45,77 +45,80 @@ impl<T> WithPrecedence<T> {
 /// example, `if cond then.x else.y` is parsed as `if cond then .x else .y`, which after expansion
 /// translates to `if cond then (\t -> t.x) else (\t -> t.y)`. However, for some macros spacing is
 /// not needed. For example, `(.x)` is parsed as `(\t -> t.x)`, which is understandable.
-fn annotate_tokens_that_need_spacing(items: Vec<syntax::Item>) -> Vec<syntax::Item> {
-    // TODO: It should be possible to make it faster by iterating over mut vec. To be checked.
-    items
-        .into_iter()
-        .map(|item| match item {
-            syntax::Item::Block(_) => item,
-            syntax::Item::Token(_) => item,
-            syntax::Item::Tree(ast) => syntax::Item::Tree(match &*ast.variant {
-                syntax::tree::Variant::MultiSegmentApp(data)
-                    if !data.segments.first().header.is_symbol() =>
-                    ast.with_error("This expression cannot be used in a non-spaced equation."),
-                _ => ast,
-            }),
-        })
-        .collect()
+fn annotate_tokens_that_need_spacing(item: syntax::Item) -> syntax::Item {
+    use syntax::tree::Variant::*;
+    item.map_tree(|ast| match &*ast.variant {
+        MultiSegmentApp(data) if !data.segments.first().header.is_symbol() =>
+            ast.with_error("This expression cannot be used in a non-spaced equation."),
+        _ => ast,
+    })
 }
 
-/// Take [`Item`] stream, resolve operators precedence and return the final AST. The precedence
-/// resolution algorithm bases on the [Shunting yard algorithm](https://en.wikipedia.org/wiki/Shunting_yard_algorithm).
-/// It is extended to handle operator sections.
-#[inline(always)]
+/// Take [`Item`] stream, resolve operator precedence and return the final AST.
+///
+/// The precedence resolution algorithm is based on the Shunting yard algorithm[1], extended to
+/// handle operator sections.
+/// [1]: https://en.wikipedia.org/wiki/Shunting_yard_algorithm
 pub fn resolve_operator_precedence<'s>(items: Vec<syntax::Item<'s>>) -> syntax::Tree<'s> {
+    // If `items` is empty, opt_rhs.unwrap() will panic below. Let's panic now with a clear error.
+    assert!(!items.is_empty());
```

Comment:
We should remove all possible panics from the code. Maybe we should accept `NonEmptyVec` instead here and move the possible error handling to the call side?

Reply:
I've been thinking about doing just that to avoid this (and the `opt_rhs` panic below, and all the `if let` branches that should *always* match). However, the way the shunting yard algorithm shuffles values around, it's not trivial to prove to the typechecker the relevant property: *If the input is non empty, we will have at least one output*, which allows us to construct a return value. (If we don't prove it to the typechecker, we need to unwrap somewhere--or return an `Option`, but some callers would need to unwrap that). I have a way to do it in mind (and eliminate *all* the branches in this function that should never be taken), but I'll need to totally rewrite this function, so I've been planning on doing it later.

Reply:
Why don't we accept NonEmptyVec and we convert it to vec just in the first line and then we will leave the panic in the code with comment that its unreachable because there is always at least one arg?